### PR TITLE
GEODE-369: wait for primary to be available when going from 0 to 1 server

### PR DESCRIPTION
**Pull request by mhansonp**
Co-Authored-By: Bill Burcham <bburcham@pivotal.io>

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelTestBase.java [hunk 1]
```diff
@@ -132,380 +124,227 @@ public final void postSetUp() throws Exception {
...
-    // + deadServers, (System.currentTimeMillis() - start) < maxWaitTime);
-    // try {
-    // Thread.yield();
-    // synchronized(delayLock) {delayLock.wait(4000);}
-    // }
-    // catch (InterruptedException ie) {
-    // fail("Interrupted while waiting ", ie);
-    // }
-    // }
+    await(
```
Comment (mhansonp) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelTestBase.java]: Here is the important change for the primary problem. Elsewhere we cleaned up and normalized most of the timeouts to 3 minutes each, and used Awaitility.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelTestBase.java [hunk 2]
```diff
@@ -132,380 +124,227 @@ public final void postSetUp() throws Exception {
...
-    Wait.waitForCriterion(wc, 60 * 1000, 2000, true);
+    await()
+        .atMost(180, SECONDS).pollInterval(2, SECONDS)
+        .until(() -> pool.getRedundantNames().contains(server));
   }
 
-  public static void verifyLiveAndRedundantServers(final int liveServers,
+  /**
+   *
+   * @param connectedServers is the expected number of connected servers (0 or more)
```
Comment (Bill) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelTestBase.java]: This should say "1 or more". The big learning here was that it's not possible to test for zero connected servers via the API (calling through the API in that case results in the `NoSubscriptionServersAvailableException` exception).

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelTestBase.java [hunk 3]
```diff
@@ -132,380 +124,227 @@ public final void postSetUp() throws Exception {
...
+   * @param redundantServers is the expected number of redundant servers (0 or more)
+   */
+  static void verifyConnectedAndRedundantServers(final int connectedServers,
```
Comment (Bill) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelTestBase.java]: Should throw `IllegalArgumentException` if `connectedServers < 1` or `redundantServers < 0`

**Comment by mhansonp** [comment]
@upthewaterspout @WireBaron @balesh2 @kirklund @pdxrunner Can you take a look at this PR?

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelTestBase.java [hunk 4]
```diff
@@ -132,380 +124,231 @@ public final void postSetUp() throws Exception {
...
-    } catch (Exception ignore) {
-      // not sure why it's ok to ignore but if you don't ignore it, RedundancyLevelPart3DUnitTest
-      // will fail
-    }
+  private static void putEntriesK1andK2() throws InterruptedException {
+    Region<String, String> r1 = cache.getRegion(Region.SEPARATOR + REGION_NAME);
+    assertThat(r1).isNotNull();
+    r1.put(k1, k1);
+    r1.put(k2, k2);
+    assertThat(r1.get(k1)).isEqualTo(k1);
```
Comment (mhansonp) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelTestBase.java]: shifting to get from get entry was the fix here. The rest is refactoring.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelTestBase.java [hunk 5]
```diff
@@ -132,380 +124,231 @@ public final void postSetUp() throws Exception {
...
-      Wait.waitForCriterion(wc, 180 * 1000, 2000, true);
+    // one client is connected to this server
+    final CacheClientNotifier ccn = bs.getAcceptor().getCacheClientNotifier();
+    await().atMost(180, SECONDS).pollInterval(1, SECONDS).until(() -> ccn.getClientProxies().size(),
+        equalTo(1));
+  }
 
-      Iterator iter_prox = ccn.getClientProxies().iterator();
+  static void verifyInterestRegistration() {
+    await("Number of bridge servers (" + cache.getCacheServers().size() + ") never became 1")
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelTestBase.java]: I would delete all usage of pollInterval and then crank up the atMost to FIVE MINUTES.

Comment (mhansonp) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelTestBase.java]: sometimes, I don't want the system to poll more frequently than specified. Thats why I use pollInterval. Do you think that is a bad approach? I thought the default was 100ms

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelPart1DUnitTest.java [hunk 6]
```diff
@@ -461,40 +297,25 @@ public void testRedundancySpecifiedNonPrimaryEPFailsDetectionByUnregisterInteres
...
+    createClientCache(NetworkUtils.getServerHostName(), PORT1, PORT2, PORT3, PORT4,
+        1, 250, 500);
+    waitConnectedServers(4);
+    assertThat(pool.getRedundantNames().size()).isEqualTo(1);
+    assertThat(SERVER1).isEqualTo(pool.getPrimaryName());
+    assertThat(pool.getRedundantNames().contains(SERVER2)).isTrue();
+    verifyOrderOfEndpoints();
+    server1.invoke(RedundancyLevelTestBase::stopServer);
+    doPuts();
+    System.out.println("server1=" + SERVER1);
```
[the remark this answers is not in the document]
Comment (mhansonp) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelPart1DUnitTest.java]: done

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelPart1DUnitTest.java [hunk 7]
```diff
@@ -42,99 +39,54 @@ public static void caseSetUp() throws Exception {
...
-      }
-
-      public String description() {
-        return "Connected server count (" + pool.getConnectedServerCount() + ") never became "
-            + expected;
-      }
-    };
-    Wait.waitForCriterion(wc, 60 * 1000, 1000, true);
+    await("Connected server count (" + pool.getConnectedServerCount() + ") never became "
+        + expected).atMost(60, SECONDS).pollInterval(1, SECONDS)
```
[the remark this answers is not in the document]
Comment (mhansonp) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelPart1DUnitTest.java]: done

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelPart1DUnitTest.java [hunk 8]
```diff
@@ -42,99 +39,54 @@ public static void caseSetUp() throws Exception {
...
-      assertEquals(SERVER2, pool.getPrimaryName());
-    } catch (Exception ex) {
-      Assert.fail("test failed due to exception in test RedundancyNotSpecifiedPrimaryServerFails ",
-          ex);
-    } /*
-       * finally { ClientServerObserverHolder.setInstance(oldBo); }
-       */
+  public void testRedundancyNotSpecifiedPrimaryServerFails() throws Exception {
+
+    // Asif: Increased the socket read timeout to 3000 sec because the registering
```
[the remark this answers is not in the document]
Comment (mhansonp) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelPart1DUnitTest.java]: done

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelTestBase.java [hunk 9]
```diff
@@ -132,380 +124,231 @@ public final void postSetUp() throws Exception {
...
-    }
+  public static void startServer() throws IOException {
+    Cache c = CacheFactory.getAnyInstance();
+    CacheServerImpl bs = (CacheServerImpl) c.getCacheServers().iterator().next();
+    assertThat(bs).isNotNull();
+    bs.start();
   }
 
-  private void createCache(Properties props) throws Exception {
+  private void createCache(Properties props) {
```
[the remark this answers is not in the document]
Comment (mhansonp) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelTestBase.java]: done

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelTestBase.java [hunk 10]
```diff
@@ -531,80 +374,61 @@ public void beforeFailoverByCacheClientUpdater(ServerLocation epFailed) {
...
-    factory.setMirrorType(MirrorType.KEYS_VALUES);
-    RegionAttributes attrs = factory.createRegionAttributes();
+    factory.setEnableSubscriptionConflation(true);
+    factory.setDataPolicy(DataPolicy.REPLICATE);
+    RegionAttributes<String, String> attrs = factory.createRegionAttributes();
     cache.createVMRegion(REGION_NAME, attrs);
 
     CacheServer server1 = cache.addCacheServer();
 
     int port = AvailablePort.getRandomAvailablePort(AvailablePort.SOCKET);
```
[the remark this answers is not in the document]
Comment (mhansonp) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelTestBase.java]: done

**Comment by mhansonp** [comment]
@upthewaterspout @kirklund  I have reviewed Kirk's concerns and addressed them I believe. The failure in Build was in an unrelated are of the code.  Please merge if you get the opportunity.

**Pull request merged**